src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs

```diff
@@ -323,7 +324,7 @@ public static DependencyGraphSpec Load(string path)
                                 {
                                     PackageSpec packageSpec = JsonPackageSpecReader.GetPackageSpec(jsonReader, path);
 
-                                    dgspec._projects.Add(projectsPropertyName, packageSpec);
+                                    dgspec._projects.TryAdd(projectsPropertyName, packageSpec);
```

Comment:
The semantics here are probably changing. I'm not sure about `SortedDictionary<TKey, TValue>`, but `Dictionary<TKey, TValue>` throws when you try to call `.Add` on a key that already exists. The new code will silently ignore the new value.

Reply:
Changed the syntax.

Reply:
I added wrapper function check duplicate key value before add value.

> This PR is less effort than my preferred option, but I'd prefer a more "functional" (less side-effects) approach. In my opinion, DependencyGraphSpec should be immutable, and the parallel foreach should not mutate the object for each package, but should do create the PackageSpec in parallel, then use the orchestrating task/thread to accumulate the results in a lock-free way. Very roughly:
> 
> ```cs
> var tasks = new List<Task<PackageSpec>>();
> var packageSpecs = List<PackageSpec>();
> IEnumerable<Project> enumerator = projects.GetEnumerable();
> 
> for (int i = 0; i < Environment.ProcessorCount; i++)
> {
>   if (!enumerator.MoveNext())
>   {
>     break;
>   }
>   tasks.Add(Task.Run(() => GetProjectSpec(enumerator.Current)));
> }
> 
> while (tasks.Count > 0)
> {
>   var finished = await Task.WhenAny(tasks);
>   tasks.Remove(finished);
>   packageSpecs.Add(finished.Result);
> 
>   if (enumerator.Current != null && enumerator.MoveNext())
>   {
>     tasks.Add(() => GetPackageSpec(enumerator.Current));
>   }
> }
> 
> var dgSpec = new DependencyGraphSpec(packageSpecs);
> ```
> 
> Even better would be to use TPL dataflow, rather than building our own concurrency limiter. In one of Martin's PRs I already provided sample code on how the same thing is achieveable with TPL dataflow. I remember that NuGet.Build.Tasks.Console already uses TPL dataflow, so it should be very easy to use there. Making the DGSpec immutable would be much less easy.

I'll take into my other PR.